src/charm.py

```diff
@@ -356,6 +356,31 @@ def _block_on_openstack_config(self, state: CharmState) -> bool:
             return True
         return False
 
+    def _common_install_code(self, state: CharmState) -> bool:
+        """Installation code shared between install and upgrade hook.
+
+        Returns:
+            Whether the installation code was successful.
+        """
+        self.unit.status = MaintenanceStatus("Installing packages")
+        try:
+            # The `_start_services`, `_install_deps` includes retry.
+            self._install_deps()
```

Comment:
Out of the scope of this implementation but I'm wondering if the way we do apt-installs can use the [apt lib for machine charms](https://charmhub.io/operator-libs-linux/libraries/apt) rather than doing subprocess commands. This might be beneficial so that we can keep the SubprocessError separate to custom execution commands rather than apt-install and command executions.

There is a separate ticket for it (ISD-1394) so no need to do it here to keep the scope!